usr/src/uts/common/fs/zfs/sys/spa.h

```diff
@@ -326,7 +327,8 @@ typedef enum bp_embedded_type {
 typedef struct blkptr {
 	dva_t		blk_dva[SPA_DVAS_PER_BP]; /* Data Virtual Addresses */
 	uint64_t	blk_prop;	/* size, compression, type, etc	    */
-	uint64_t	blk_pad[2];	/* Extra space for the future	    */
+	uint32_t	blk_iv[3];	/* IV for block level encryption    */
+	uint32_t	blk_pad;	/* Extra space for the future	    */
```

Comment:
Unfortunately, not really. The choices seem to be:

1) DVA[2], which will interfere with the Nexenta feature and reduce the max possible data copies. I feel this might not really make sense to users.
2)The last 128 bits of the checksum, which will reduce our checksum strength. In fact, at this point we won't have a traditional checksum at all (we will have the 128 bit MAC from AES, however).
3) The padding where it currently resides.

I will say this: Theoretically, I think it should be possibly to use `BP_IDENTITY()` to be able to generate an IV at runtime for both reads and writes in the normal case. However, with dedup on, this method falls apart because the following will occur:

1) A block `x`, which should match block `y` for dedup, will get allocated to some spot on disk
2) This spot on disk will be different from block `y`'s, so the IV will be different
3) The encryption algorithm will spit out a different MAC for the block, since a different IV was given
4) The checksum will be generated, which will also be different because the cipher data will be different due to the different IV.
4) Dedup will not happen both because the checksum is wrong and the MAC is wrong (which also resides with part of the checksum).

So maybe a "crappy, but not as crappy as the alternatives" solution is to just make dedup and encryption mutually exclusive. I just thought of all of this now, so there might be some flaws in the idea that make it unfeasible / insecure. But maybe someone else can give me their thoughts?